Inconsistent date parsing of to_datetime

MarcoGorelli opened the pull request:
- [x] xref #12585
- [x] tests added / passed
- [ ] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [ ] whatsnew entry

carries on from https://github.com/pandas-dev/pandas/pull/35428

arw2019 committed:
added warnings when parse inconsistent with dayfirst arg

arw2019 committed:
improved error message

arw2019 committed:
TST: added tests

arw2019 committed:
removed trailing whitespaces

arw2019 committed:
removed pytest.warns

MarcoGorelli committed:
wip

MarcoGorelli committed:
revert

MarcoGorelli committed:
set stacklevel, assert warning messages

jreback commented:
@MarcoGorelli this looks neat. prob need to troll all of the issues and add every test you can find :->

MarcoGorelli committed:
okwarning in user guide

MarcoGorelli committed:
:art:

MarcoGorelli committed:
catch warnings

MarcoGorelli committed:
fixup

mroeschke reviewed (commented):
Nice looks good. Just needs a whatsnew note for 1.4 (probably good to have a dedicated section demoing this behavior), and maybe the `to_datetime` docstring can explain when a warning will be raised.

MarcoGorelli committed:
add to to_datetime docstring, add whatsnew note

MarcoGorelli commented:
Sure, added, here's the docstring:

![image](https://user-images.githubusercontent.com/33491632/128625774-b9f8816c-323d-4be6-8a18-dfccddeb3f49.png)

Unfortunately, this doesn't solve the case of datetime strings, just date strings. E.g.:
```python
In [2]: pd.to_datetime(['31-12-2021'])
<ipython-input-2-a6d9926683b2>:1: UserWarning: Parsing '31-12-2021' in DD/MM/YYYY format. Provide format or specify infer_datetime_format=True for consistent parsing.
  pd.to_datetime(['31-12-2021'])
Out[2]: DatetimeIndex(['2021-12-31'], dtype='datetime64[ns]', freq=None)

In [3]: pd.to_datetime(['31-12-2021 12:00:31'])
Out[3]: DatetimeIndex(['2021-12-31 12:00:31'], dtype='datetime64[ns]', freq=None)
```
because those cases go straight to dateutil:

https://github.com/pandas-dev/pandas/blob/08d296f1278e08b407448c95086589e1d10285f9/pandas/_libs/tslibs/parsing.pyx#L238-L244

and this PR only addresses what reaches `_parse_delimited_date`.

> Just needs a whatsnew note for 1.4 (probably good to have a dedicated section demoing this behavior),

As in, in the "notable bug fixes" section? If so, I've added it there, here's how it looks:

![image](https://user-images.githubusercontent.com/33491632/128626108-339ab67c-bd0e-47b1-be9f-86c0b12a995e.png)

jreback commented:
https://github.com/pandas-dev/pandas/blob/08d296f1278e08b407448c95086589e1d10285f9/pandas/_libs/tslibs/parsing.pyx#L238-L244

yeah would be nice to warn in this as well

MarcoGorelli commented:
> yeah would be nice to warn in this as well

Agreed, just not sure how to do it - all I can think of is

```python
        dt = du_parse(date_string, default=_DEFAULT_DATETIME,
                      dayfirst=dayfirst, yearfirst=yearfirst, **kwargs)
        if dayfirst and not re.search(rf'(?<!\d){dt.day}(?!\d)', date_string).start() < re.search(rf'(?<!\d){dt.month}(?!\d)', date_string).start():
            warnings.warn(
                PARSING_WARNING_MSG.format(
                    date_string=date_string,
                    format='MM/DD/YYYY'
                ),
                stacklevel=4,
            )
        elif not dayfirst and re.search(rf'(?<!\d){dt.day}(?!\d)', date_string).start() < re.search(rf'(?<!\d){dt.month}(?!\d)', date_string).start():
            warnings.warn(
                PARSING_WARNING_MSG.format(
                    date_string=date_string,
                    format='DD/MM/YYYY'
                ),
                stacklevel=4,
            )
```
would that be OK?

jreback commented:
well the bigger problem here is that we *really* should warn anytime this is called `du_parse` (if we are not guessing a format) and actually parsing, e.g. this is the fallback case.

I am happy to merge this and to do followups, but ideally want to warn if at all possible that something is amiss (even with false positives).

jreback commented on doc/source/whatsnew/v1.4.0.rst:
```diff
@@ -50,10 +50,19 @@ Notable bug fixes
 
-notable_bug_fix1
-^^^^^^^^^^^^^^^^
+Inconsistent date string parsing
```
can you also add to the documentation itself, e.g. somewhere in `timeseries.rst` is appropriate

jreback commented on pandas/core/tools/datetimes.py:
```diff
@@ -701,8 +701,15 @@ def to_datetime(
+        .. warning::
+
+            dayfirst=True is not strict, but will prefer to parse
+            with day first (this is a known bug, based on dateutil behavior).
```
i wouldn't say this is a known bug unless you can point to an authoritative reference.

MarcoGorelli replied:
I didn't add this, it's from https://github.com/pandas-dev/pandas/pull/7599 , I just put it into a warning block

jreback commented:
does this show any warnings during csv parsing? (i think it should in the same cases). ideally create a followon issue with checkboxes for the cases.

jreback commented:
@MarcoGorelli looks pretty good if you can rebase & address comments.

MarcoGorelli committed:
Merge remote-tracking branch 'upstream/master' into pr/arw2019/to_datetime-inconsistent-parsing

MarcoGorelli committed:
wip

MarcoGorelli committed:
wip

MarcoGorelli committed:
wip

MarcoGorelli committed:
wip

MarcoGorelli committed:
fixup test

MarcoGorelli committed:
more fixups

MarcoGorelli committed:
fixup

MarcoGorelli committed:
revert to b4bb5b330ad25c7dbca36fe55d4c264ec4d027d1

MarcoGorelli committed:
document in timeseries.rst

MarcoGorelli committed:
add tests for read_csv

MarcoGorelli committed:
check expected_inconsistent in tests

MarcoGorelli commented:
> does this show any warnings during csv parsing? (i think it should in the same cases).

yup, I've added the same suite of tests for that

> well the bigger problem here is that we really should warn anytime this is called du_parse (if we are not guessing a format) and actually parsing, e.g. this is the fallback case.

I gave that a go but it's a lot more complicated and there's a ton more warnings to catch. I've pushed that work to a branch (MarcoGorelli:warn-on-du-parse), but for now I've kept this PR to just the delimited date case

MarcoGorelli committed:
fixup docs

MarcoGorelli committed:
Merge remote-tracking branch 'upstream/master' into pr/arw2019/to_datetime-inconsistent-parsing

mroeschke reviewed (commented):
Just a small comment about removing the comment about the dateutil bug and then feel free to merge.

MarcoGorelli committed:
remove note about dateutil bug
